homeassistant/components/yeelight/__init__.py

```diff
@@ -342,42 +339,86 @@ def __init__(self, hass: HomeAssistant) -> None:
         self._unique_id_capabilities = {}
         self._host_capabilities = {}
         self._track_interval = None
-        self._listener = None
-        self._connected_event = None
+        self._listeners = []
+        self._connected_events = []
 
     async def async_setup(self):
         """Set up the scanner."""
-        if self._connected_event:
-            await self._connected_event.wait()
+        if self._connected_events:
+            await asyncio.gather(*(event.wait() for event in self._connected_events))
             return
-        self._connected_event = asyncio.Event()
 
-        async def _async_connected():
-            self._listener.async_search()
-            self._connected_event.set()
+        for idx, source_ip in enumerate(await self._async_build_source_set()):
+            self._connected_events.append(asyncio.Event())
+
+            def _wrap_async_connected_idx(idx):
+                """Create a function to capture the idx cell variable."""
+
+                async def _async_connected():
+                    self._connected_events[idx].set()
+
+                return _async_connected
+
+            self._listeners.append(
+                SSDPListener(
+                    async_callback=self._async_process_entry,
+                    service_type=SSDP_ST,
+                    target=SSDP_TARGET,
+                    source_ip=source_ip,
+                    async_connect_callback=_wrap_async_connected_idx(idx),
+                )
+            )
 
-        self._listener = SSDPListener(
-            async_callback=self._async_process_entry,
-            service_type=SSDP_ST,
-            target=SSDP_TARGET,
-            async_connect_callback=_async_connected,
+        results = await asyncio.gather(
+            *(listener.async_start() for listener in self._listeners),
+            return_exceptions=True,
         )
-        await self._listener.async_start()
-        await self._connected_event.wait()
+        failed_listeners = []
+        for idx, result in enumerate(results):
+            if not isinstance(result, Exception):
+                continue
+            _LOGGER.warning(
+                "Failed to setup listener for %s: %s",
+                self._listeners[idx].source_ip,
+                result,
+            )
+            failed_listeners.append(self._listeners[idx])
+            self._connected_events[idx].set()
+
+        for listener in failed_listeners:
+            self._listeners.remove(listener)
+
+        await asyncio.gather(*(event.wait() for event in self._connected_events))
```

Comment:
Can `self._connected_events` be empty? If so, that would break `asyncio.gather` (as it requires at least 1 task)

Reply:
If it's empty it means they can't listen on any configured address and discovery is broken either way.

I guess it's a question of where it should blow up at that point.

Reply:
Maybe logger error and tell them to check their network settings and return at that point.

Reply:
Actually it can never be empty, as we set them on the failure case to ensure we don't end up trying to connect over and over.

Reply:
We should probably add a warning when setting up a listener fails and we skip it

Reply:
Nevermind, I already did that on line 381.

Too much task switching.

I think this is good.